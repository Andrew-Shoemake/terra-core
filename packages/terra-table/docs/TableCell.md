# Terra Table Cell

The Terra Table Cell is a sub component to be used inside a table row.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table`
  - `yarn add terra-table`

## Usage

```jsx
import React from 'react';
import { Cell } from 'terra-table';

<Cell key="row-0-cell-0">
  <p>Table row text</p>
</Cell>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
