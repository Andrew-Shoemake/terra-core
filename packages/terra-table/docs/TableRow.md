# Terra Table Row

The Terra Table Row is a sub component that groups child cells. Each row handles it's own selection state through props. The onSelect function that can be passed to the table row will return the metaData associated to the row. The metaData allows for greater control of state with the ability associate any indicator or function with the row.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table`
  - `yarn add terra-table`

## Usage

```jsx
import React from 'react';
import { Row, Cell } from 'terra-table';

  <Row
    key="unique-react-key"
    isSelectable
    isSelected
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Cell>
      Cell Content
    </Cell>
    <Cell>
      Cell Content
    </Cell>
  </Row>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
