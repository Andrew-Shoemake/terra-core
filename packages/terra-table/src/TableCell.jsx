import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import styles from './TableCell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Content to be displayed for the row cell
   */
  children: PropTypes.node,
  /**
   * Function callback for the ref of the td.
   */
  refCallback: PropTypes.func,
};

const TableCell = ({
  children,
  refCallback,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
    customProps.className,
  ]);

  return (
    <td {...customProps} className={cellClassNames} ref={refCallback} role="gridcell">
      {children}
    </td>
  );
};

TableCell.propTypes = propTypes;

export default TableCell;
