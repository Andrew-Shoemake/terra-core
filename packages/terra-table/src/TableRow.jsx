import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import styles from './TableRow.module.scss';
import TableUtils from './TableUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node,
  /**
   * Whether or not row is selected
   */
  isSelected: PropTypes.bool,
  /**
   * Whether or not row is selectable
   */
  isSelectable: PropTypes.bool,
  /**
   * The associated metaData to be provided in the onSelect callback.
   */
  // eslint-disable-next-line react/forbid-prop-types
  metaData: PropTypes.object,
  /**
   * Function callback for when the appropriate click or key action is performed.
   * Callback contains the javascript evnt and prop metadata, e.g. onSelect(event, metaData)
   */
  onSelect: PropTypes.func,
  /**
   * Function callback for the ref of the tr.
   */
  refCallback: PropTypes.func,
};

const defaultProps = {
  isSelected: false,
  isSelectable: false,
};

const TableRow = ({
  children,
  isSelected,
  isSelectable,
  metaData,
  onBlur,
  onClick,
  onKeyDown,
  onMouseDown,
  onSelect,
  refCallback,
  ...customProps
}) => {
  const rowClassNames = cx([
    { 'is-selected': isSelected && isSelectable },
    { 'is-selectable': isSelectable },
    'row',
    customProps.className,
  ]);

  const attrSpread = {};
  if (isSelectable) {
    attrSpread.onClick = TableUtils.wrappedOnClickForItem(onClick, onSelect, metaData);
    attrSpread.onKeyDown = TableUtils.wrappedOnKeyDownForItem(onKeyDown, onSelect, metaData);
    attrSpread.tabIndex = '0';
    attrSpread['aria-selected'] = isSelected;
    attrSpread['data-row-show-focus'] = 'true';
    attrSpread.onBlur = TableUtils.wrappedEventCallback(onBlur, event => event.currentTarget.setAttribute('data-row-show-focus', 'true'));
    attrSpread.onMouseDown = TableUtils.wrappedEventCallback(onMouseDown, event => event.currentTarget.setAttribute('data-row-show-focus', 'false'));
  }

  return (
    <tr {...customProps} {...attrSpread} className={rowClassNames} ref={refCallback}>
      {children}
    </tr>
  );
};

TableRow.propTypes = propTypes;
TableRow.defaultProps = defaultProps;

export default TableRow;
