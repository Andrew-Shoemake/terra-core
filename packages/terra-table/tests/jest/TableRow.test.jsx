import React from 'react';
import TableRow from '../../src/TableRow';
import TableCell from '../../src/TableCell';

// Snapshot test
it('should render a default table row', () => {
  const tableRow = shallow(<TableRow />);
  expect(tableRow).toMatchSnapshot();
});

it('should render a table row with a cell', () => {
  const tableRow = shallow(<TableRow><TableCell /></TableRow>);
  expect(tableRow).toMatchSnapshot();
});

it('should render a selectable table row', () => {
  const tableRow = shallow(<TableRow isSelectable />);
  expect(tableRow).toMatchSnapshot();
});

it('should render a selected table row', () => {
  const tableRow = shallow(<TableRow isSelected />);
  expect(tableRow).toMatchSnapshot();
});
