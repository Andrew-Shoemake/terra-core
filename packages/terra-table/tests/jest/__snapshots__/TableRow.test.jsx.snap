// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default table row 1`] = `
<tr
  className="row"
/>
`;

exports[`should render a selectable table row 1`] = `
<tr
  aria-selected={false}
  className="is-selectable row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  tabIndex="0"
/>
`;

exports[`should render a selected table row 1`] = `
<tr
  aria-selected={true}
  className="is-selected is-selectable row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  tabIndex="0"
/>
`;

exports[`should render a striped table row 1`] = `
<tr
  className="is-striped row"
/>
`;

exports[`should render a table row with a cell 1`] = `
<tr
  className="row"
>
  <TableCell />
</tr>
`;

exports[`should render a table row with onSelect 1`] = `
<tr
  aria-selected={false}
  className="is-selectable row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  tabIndex="0"
/>
`;

exports[`should render a table row with refCallback 1`] = `
<tr
  className="row"
/>
`;
